### ISSUE TITLE
Electra Models have bad performance

### ISSUE THREAD

**Issue by PhilipMay** [UKPLab/sentence-transformers#362]
Hi,
this issue is in connection with #193 ...

I trained german sentence transformers on our brand new german electra model `german-nlp-group/electra-base-german-uncased` (1) (2) but have relativly bad results compared to other german models. But we showed that our new electra model is better on downstream tasks (germeval18 and 17).

I train on a german xnli dataset. `dbmdz/bert-base-german-uncased` reaches 0.731166 COSINE spearman while the electra model barely reaches 0.6 - both are optimized with optuna over several steps.

What might be the reason? Do you have a theory?
AFAIK the last layer of BERT and Electra are identical. So I have no idea why it differs so much.

(1): through a bug it does not show up on HF yet - see here https://github.com/huggingface/transformers/issues/6495
(2): model card is here: https://github.com/German-NLP-Group/german-transformer-training/blob/master/model_cards/electra-base-german-uncased.md

**Comment by nreimers** [comment]
Did you do any further fine tuning or used the Electra model out of the box?

In general better down stream performances is sadly not a reliable indicator of producing better sentence embeddings. For sentence embeddings, all dimensions are treated equally with the same weight. This is not needed for downstream task, here Irrelevant dimensions or dimensions of different scales would not harm the performance for downstream tasks.

**Comment by PhilipMay** [comment]
> Did you do any further fine tuning or used the Electra model out of the box?

As I wrote in the issue:
I did use our new german electra LM and then used sentence-transformers to train the model on a german xnli dataset.
Since electra and BERT (dbmdz/bert-base-german-uncased) are basicaly the same architectures I do not understand why electra has such a low performance. As @nreimers said here: #193 it might be a systematic problem with electra - but why?

**Comment by nreimers** [comment]
Hi @PhilipMay 
Oh, I didn't saw it.

I am not sure why this happens. This also happens with other models like XLNet / Alberta etc., for supervised down stream tasks they perform better than BERT, but they produce worse sentence embeddings even after fine-tuning.

One reason might be that BERT was trained with the Next Sentence Prediction (NSP) task. This should be quite helpful for learning good embeddings. However, RoBERTa was trained without NSP and perform on-par or even better than BERT for sentence embeddings.

The XNLI dataset is quite small for fine-tuning. It could be extended by translation the Englich SNLI+MNLI data e.g. with these EN-DE translation model (https://github.com/pytorch/fairseq/tree/master/examples/translation).
Then the sentence embeddings could be fine-tuned on the machine translated SNLI+MNLI data.

Best
Nils Reimers

**Comment by PhilipMay** [comment]
@nreimers do you suggest a two objectives training (iterated in a round-robin way) or first translated SNLI+MNLI and then sts in a 2nd step?

**Comment by nreimers** [comment]
Hi @PhilipMay 
I haven't evaluated what is better: 1) multi-task training on NLI+STS or 2) first NLI and then STS training. 

This script does 1), while I personally use 2) in my experiment. But I never evaluated what approach is better.

**Comment by PhilipMay** [comment]
@nreimers many thanks for this very detailed answer! I will evaluate 1) vs. 2).
btw: Are you interested in doing some joined research in this area?

**Comment by nreimers** [comment]
Hi @PhilipMay 
I think sampling for 1) is quite important. Out-of-the-box, with round-robin, the results are rather bad.

I would like to join, but I am afraid that I do not really have time. I am currently working on several other research projects, so I sadly don't have time for further projects. 

But looking forward to your experiences.

**Comment by PhilipMay** [comment]
Hey @nreimers - can you please have a look here:
https://huggingface.co/transformers/model_doc/electra.html

> nearly no changes done to the underlying model: BERT. The only change is the separation of the embedding size and the hidden size -> The embedding size is generally smaller, while the hidden size is larger. An additional projection layer (linear) is used to project the embeddings from their embedding size to the hidden size. 

Might that be the reason why the sentence transformers performance is not good with Electra?

Here are the changes they made in FARM to use Electra Models: https://github.com/deepset-ai/FARM/blob/master/farm/modeling/language_model.py#L1255-L1265

Might be similar changes needed here? What do you think?

**Comment by nreimers** [comment]
Hi @PhilipMay 
I don't use the pooled output of BERT.

The pooled output of BERT is used for the next sentence prediction loss when training BERT. However, BERT mean pooling as well as using the CLS token (without the pooled layer) also works quite well for BERT, but sadly not for Electra.

Best
Nils

**Comment by PhilipMay** [comment]
Well - since there is nothing to "fix" I close this.
